code/assurance/src/main/java/com/adobe/marketing/mobile/assurance/QuickConnectManager.kt

```diff
@@ -0,0 +1,213 @@
+
+        val quickConnectDeviceCreator = QuickConnectDeviceCreator(orgId, clientId, deviceName) {
+            when (it) {
+                is Response.Success -> checkDeviceStatus(orgId, clientId)
+                is Response.Failure -> quickConnectCallback.onError(it.error)
+            }
+        }
+
+        deviceCreationTaskHandle = executorService.submit(quickConnectDeviceCreator)
+    }
+
+    /**
+     * Periodically checks the status of quick connect device registration that was triggered by [registerDevice]
+     *
+     * @param orgId the orgId for which quick connect was initiated
+     * @param clientId the clientId for which quick connect was initiated
+     */
+    @VisibleForTesting
+    internal fun checkDeviceStatus(orgId: String, clientId: String) {
+        val statusCheckerTask = QuickConnectDeviceStatusChecker(orgId, clientId) { response ->
+            handleStatusCheckResponse(orgId, clientId, response)
+        }
+
+        deviceStatusTaskHandle = executorService.schedule(statusCheckerTask, QuickConnect.STATUS_CHECK_DELAY_MS, TimeUnit.MILLISECONDS)
+    }
+
+    /**
+     * Cancels an ongoing quick connect device registration workflow (if any).
+     */
+    internal fun cancel() {
+        cleanup()
+    }
+
+    /**
+     * Handles the response from the device status check. Conditionally triggers a new status check
+     * if the request was successful without session details.
+     *
+     * @param orgId the orgId for which quick connect was initiated
+     * @param clientId the clientId for which quick connect was initiated
+     * @param response the [Response] from the [checkDeviceStatus] request that is to be handled
+     */
+    private fun handleStatusCheckResponse(orgId: String, clientId: String, response: Response<HttpConnecting, AssuranceQuickConnectError>) {
+        when (response) {
+            is Response.Success -> {
+                val sessionDetails = extractSessionDetails(StreamUtils.readAsString(response.data.inputStream))
+                if (sessionDetails != null) {
+                    // quick connect session details are available. Notify about successful the result.
+                    Log.trace(LOG_TAG, LOG_SOURCE, "Received session details.")
+
+                    quickConnectCallback.onSuccess(sessionDetails.sessionId, sessionDetails.token)
+                    cleanup()
+                    return
+                }
+
+                // The request was successful but the session data is not yet present.
+
+                if (!isActive) {
+                    // The workflow is likely cancelled due to user interaction. Do not retry.
+                    Log.trace(LOG_TAG, LOG_SOURCE, "Will not retry. QuickConnect workflow already cancelled.")
+                    return
+                }
+
+                if (++retryCount < QuickConnect.MAX_RETRY_COUNT) {
```

Comment:
Given that MAX_RETRY_COUNT is 300 and the timeout is 2000ms, is it really necessary that we wait 10 minutes to see if QuickConnect can/is established?

Reply:
I agree 😅. 10 mins is indeed long for waiting on the user to login to the Assurance UI and selecting their device there. I can sync with product about it - do you have any suggestions on what would be a good time out given that we will also have a retry button on the UI incase the status fetch fails?

Reply:
I think it's fine for now, not too worried about it.  It makes sense that it's longer because of the waiting for an interaction on another system, but 10 minutes does seem a touch excessive :).